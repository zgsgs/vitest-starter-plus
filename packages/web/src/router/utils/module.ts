/**
 * 权限路由排序
 * @param routes - 权限路由
 */
export function sortRoutes(routes: AuthRoute.Route[]) {
  return routes
    .sort((next, pre) => Number(next.meta?.order) - Number(pre.meta?.order))
    .map((i) => {
      if (i.children)
        sortRoutes(i.children)
      return i
    })
}

/**
 * 处理全部导入的路由模块
 * @param modules - 路由模块
 */
export function handleModuleRoutes(modules: AuthRoute.RouteModule) {
  const routes = Object.keys(modules).reduce((acc: AuthRoute.Route[], key) => {
    const item = modules[key].default
    if (item)
      acc.push(item)
    else
      window.console.error(`路由模块解析出错: key = ${key}`)

    return acc
  }, [])

  return sortRoutes(routes)
}
