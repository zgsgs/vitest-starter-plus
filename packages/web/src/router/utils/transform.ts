import type { RouteRecordRaw } from 'vue-router'
import { getLayoutComponent, getViewComponent } from './component'

/** 将权限路由转换成vue路由
 * @param routes - 权限路由
 * @description 所有多级路由都会被转换成二级路由
 */
export function transformAuthRouteToVueRoutes(routes: AuthRoute.Route[]) {
  return routes.map(route => transformAuthRouteToVueRoute(route)).flat(1)
}

type ComponentAction = Record<AuthRoute.RouteComponentType, () => void>

/** 将单个权限路由转换成vue路由
 * @param route - 单个权限路由
 */
export function transformAuthRouteToVueRoute(route: AuthRoute.Route) {
  const resultRoute: RouteRecordRaw[] = []

  const itemRoute = { ...route } as RouteRecordRaw

  // 动态path
  if (hasDynamicPath(route))
    Object.assign(itemRoute, { path: route.meta.dynamicPath })

  // 外链路由
  if (hasHref(route))
    Object.assign(itemRoute, { component: getViewComponent('404') })

  // 路由组件
  if (hasComponent(route)) {
    const action: ComponentAction = {
      basic() {
        itemRoute.component = getLayoutComponent('basic')
      },
      blank() {
        itemRoute.component = getLayoutComponent('blank')
      },
      multi() {
        // 多级路由一定有子路由
        if (hasChildren(route)) {
          Object.assign(itemRoute, { meta: { ...itemRoute.meta, multi: true } })
          delete itemRoute.component
        }
        else {
          window.console.error('多级路由缺少子路由: ', route)
        }
      },
      self() {
        itemRoute.component = getViewComponent(route.name as AuthRoute.LastDegreeRouteKey)
      },
    }
    try {
      if (route.component)
        action[route.component]()

      else
        window.console.error('路由组件解析失败: ', route)
    }
    catch {
      window.console.error('路由组件解析失败: ', route)
    }
  }

  // 注意：单独路由没有children
  if (isSingleRoute(route)) {
    if (hasChildren(route))
      window.console.error('单独路由不应该有子路由: ', route)

    // 捕获无效路由的需特殊处理
    if (route.name === 'not-found') {
      itemRoute.children = [
        {
          path: '',
          name: route.name,
          component: getViewComponent('not-found'),
        },
      ]
    }
    else {
      const parentPath = `${itemRoute.path}-parent` as AuthRouteUtils.SingleRouteKey

      const layout = route.meta.singleLayout === 'basic' ? getLayoutComponent('basic') : getLayoutComponent('blank')

      const parentRoute: RouteRecordRaw = {
        path: parentPath,
        component: layout,
        redirect: route.path,
        children: [itemRoute],
      }

      return [parentRoute]
    }
  }

  // 子路由
  if (hasChildren(route)) {
    const children = (route.children as AuthRoute.Route[]).map(child => transformAuthRouteToVueRoute(child)).flat()

    // 找出第一个不为多级路由中间级的子路由路径作为重定向路径
    const redirectPath = (children.find(v => !v.meta?.multi)?.path || '/') as AuthRoute.RoutePath

    if (redirectPath === '/')
      window.console.error('该多级路由没有有效的子路径', route)

    if (route.component === 'multi') {
      // 多级路由，将子路由提取出来变成同级
      resultRoute.push(...children)
      delete itemRoute.children
    }
    else {
      itemRoute.children = children
    }
    itemRoute.redirect = redirectPath
  }

  resultRoute.push(itemRoute)

  return resultRoute
}

/** 将权限路由转换成搜索的菜单数据
 * @param routes - 权限路由
 * @param treeMap
 */
export function transformAuthRouteToSearchMenus(routes: AuthRoute.Route[], treeMap: AuthRoute.Route[] = []) {
  if (routes && routes.length === 0)
    return []
  return routes.reduce((acc, cur) => {
    if (!cur.meta?.hide)
      acc.push(cur)

    if (cur.children && cur.children.length > 0)
      transformAuthRouteToSearchMenus(cur.children, treeMap)

    return acc
  }, treeMap)
}

/** 将路由名字转换成路由路径 */
export function transformRouteNameToRoutePath(name: Exclude<AuthRoute.AllRouteKey, 'not-found'>): AuthRoute.RoutePath {
  const rootPath: AuthRoute.RoutePath = '/'
  if (name === 'root')
    return rootPath

  const splitMark = '_'
  const pathSplitMark = '/'
  const path = name.split(splitMark).join(pathSplitMark)

  return (pathSplitMark + path) as AuthRoute.RoutePath
}

/** 将路由路径转换成路由名字 */
export function transformRoutePathToRouteName<K extends AuthRoute.RoutePath>(path: K) {
  if (path === '/')
    return 'root'

  const pathSplitMark = '/'
  const routeSplitMark = '_'

  const name = path.split(pathSplitMark).slice(1).join(routeSplitMark) as AuthRoute.AllRouteKey

  return name
}

/** 是否有外链
 * @param item - 权限路由
 */
function hasHref(item: AuthRoute.Route) {
  return Boolean(item.meta.href)
}

/** 是否有动态路由path
 * @param item - 权限路由
 */
function hasDynamicPath(item: AuthRoute.Route) {
  return Boolean(item.meta.dynamicPath)
}

/** 是否有路由组件
 * @param item - 权限路由
 */
function hasComponent(item: AuthRoute.Route) {
  return Boolean(item.component)
}

/** 是否有子路由
 * @param item - 权限路由
 */
function hasChildren(item: AuthRoute.Route) {
  return Boolean(item.children && item.children.length)
}

/** 是否是单层级路由
 * @param item - 权限路由
 */
function isSingleRoute(item: AuthRoute.Route) {
  return Boolean(item.meta.singleLayout)
}
