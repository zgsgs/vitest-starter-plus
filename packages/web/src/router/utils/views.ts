import type { RouteComponent } from 'vue-router'

export const views: Record<
  PageRoute.LastDegreeRouteKey,
  RouteComponent | (() => Promise<{ default: RouteComponent }>)
> = {
  '403': () => import('@/views/_builtin/403'),
  '404': () => import('@/views/_builtin/404'),
  '500': () => import('@/views/_builtin/500'),
  'not-found': () => import('@/views/_builtin/not-found'),
  'constant-page': () => import('@/views/_builtin/constant-page'),
  'login': () => import('@/views/_builtin/login/index.vue'),
  'about': () => import('@/views/about/index.vue'),
  'auth-demo_permission': () => import('@/views/about/index.vue'),
  'auth-demo_super': () => import('@/views/about/index.vue'),
  'component_button': () => import('@/views/about/index.vue'),
  'component_card': () => import('@/views/about/index.vue'),
  'component_table': () => import('@/views/about/index.vue'),
  'dashboard_analysis': () => import('@/views/about/index.vue'),
  'dashboard_workbench': () => import('@/views/about/index.vue'),
  'document_naive': () => import('@/views/about/index.vue'),
  'document_project-link': () => import('@/views/about/index.vue'),
  'document_project': () => import('@/views/about/index.vue'),
  'document_vite': () => import('@/views/about/index.vue'),
  'document_vue': () => import('@/views/about/index.vue'),
  'exception_403': () => import('@/views/about/index.vue'),
  'exception_404': () => import('@/views/about/index.vue'),
  'exception_500': () => import('@/views/about/index.vue'),
  'function_tab-detail': () => import('@/views/about/index.vue'),
  'function_tab-multi-detail': () => import('@/views/about/index.vue'),
  'function_tab': () => import('@/views/about/index.vue'),
  'function_websocket': () => import('@/views/about/index.vue'),
  'management_auth': () => import('@/views/about/index.vue'),
  'management_role': () => import('@/views/about/index.vue'),
  'management_route': () => import('@/views/about/index.vue'),
  'management_user': () => import('@/views/about/index.vue'),
  'multi-menu_first_second-new_third': () => import('@/views/about/index.vue'),
  'multi-menu_first_second': () => import('@/views/about/index.vue'),
  'plugin_charts_antv': () => import('@/views/about/index.vue'),
  'plugin_charts_echarts': () => import('@/views/about/index.vue'),
  'plugin_copy': () => import('@/views/about/index.vue'),
  'plugin_editor_markdown': () => import('@/views/about/index.vue'),
  'plugin_editor_quill': () => import('@/views/about/index.vue'),
  'plugin_icon': () => import('@/views/about/index.vue'),
  'plugin_map': () => import('@/views/about/index.vue'),
  'plugin_print': () => import('@/views/about/index.vue'),
  'plugin_swiper': () => import('@/views/about/index.vue'),
  'plugin_video': () => import('@/views/about/index.vue'),
}
