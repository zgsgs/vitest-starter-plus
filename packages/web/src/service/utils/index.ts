export * from './transform'
export * from './error'
export * from './handler'
export * from './token'
export * from './response'
