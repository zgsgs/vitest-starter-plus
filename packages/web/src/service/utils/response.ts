import type { AxiosInstance, AxiosRequestConfig } from 'axios'
import type { RequestMethod } from '../request/type'

/** 获取请求结果 统一响应格式 */
export async function getRequestResponse<T>(params: {
  instance: AxiosInstance
  method: RequestMethod
  url: string
  data?: any
  axiosConfig?: AxiosRequestConfig
}): Promise<Service.RequestResult<T>> {
  const { instance, method, url, data, axiosConfig } = params

  if (['get', 'delete'].includes(method))
    return await instance[method](url, axiosConfig)

  return await instance[method](url, data, axiosConfig)
}
