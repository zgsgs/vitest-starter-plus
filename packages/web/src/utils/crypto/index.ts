import * as CryptoJS from 'crypto-js'

/** 加密密钥 */
const CryptoSecret = '__CryptoJS_Secret__'

/**
 * encrypt the data
 * @param data 明文
 * @returns 密文
 */
export function encrypto(data: unknown) {
  const newData = JSON.stringify(data)
  return CryptoJS.AES.encrypt(newData, CryptoSecret).toString()
}

/**
 * 解密数据
 * @param cipherText 密文
 * @returns 明文
 */
export function decrypto(cipherText: string) {
  const bytes = CryptoJS.AES.decrypt(cipherText, CryptoSecret)
  const originalText = bytes.toString(CryptoJS.enc.Utf8)
  if (originalText)
    return JSON.parse(originalText)

  return null
}
