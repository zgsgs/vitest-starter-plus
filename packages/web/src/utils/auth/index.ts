import { localStg } from '@/utils'

/** 获取token */
export function getToken() {
  return localStg.get('token') || ''
}

/** 是否登录 */
export const isLogin = Boolean(getToken())

/** 获取用户信息 */
export function getUserInfo() {
  const emptyInfo: Auth.UserInfo = {
    userId: '',
    userName: '',
    userRole: 'user',
  }
  const userInfo: Auth.UserInfo = localStg.get('userInfo') || emptyInfo

  return userInfo
}

/** 去除用户相关缓存 */
export function clearAuthStorage() {
  localStg.remove('token')
  localStg.remove('refreshToken')
  localStg.remove('userInfo')
}
