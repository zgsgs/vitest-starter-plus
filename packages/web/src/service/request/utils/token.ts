import type { AxiosRequestConfig } from 'axios'
import { fetchUpdateToken } from '@/service'
import { localStg } from '@/utils'

// import { useAuth } from '@/router/hooks'

/**
 * 刷新token
 * @param axiosConfig - token失效时的请求配置
 */
export async function handleRefreshToken(axiosConfig: AxiosRequestConfig) {
  // const { resetAuth } = useAuth()
  const refreshToken = localStg.get('refreshToken') || ''
  const { data } = await fetchUpdateToken(refreshToken)
  if (data) {
    localStg.set('token', data.token)
    localStg.set('refreshToken', data.refreshToken)

    const config = { ...axiosConfig }
    if (config.headers)
      config.headers.Authorization = data.token

    return config
  }

  // resetAuth()
  return null
}
