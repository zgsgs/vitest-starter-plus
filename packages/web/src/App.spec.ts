import { mount } from '@vue/test-utils'
import { sleep } from '../test/utils'
import App from './App.vue'

describe('App.vue', () => {
  beforeEach(() => {
    // eslint-disable-next-line no-console
    console.log(11)
  })
  it('should render success', async () => {
    // given
    const wrapper = mount(App)
    // when
    await sleep(200)

    const text = wrapper.text()
    // then
    expect(text).contain('Vite + Vue + TS')
  })

  it('should 复制按钮是否存在', async () => {
    // given
    const wrapper = mount(App)
    const copyBtn = wrapper.find('.copy')
    // when
    // 触发点击事件
    // copyBtn.trigger('click')
    const isExists = copyBtn.exists()

    // then
    // 断言该按钮是否存在
    expect(isExists).toBeTruthy()
  })
  describe('should 测试', () => {
    it('should 测试', async () => {
      // given
      const text = ''
      // when
      const result = text
      // then
      expect(result).toBe(text)
    })
  })
})
