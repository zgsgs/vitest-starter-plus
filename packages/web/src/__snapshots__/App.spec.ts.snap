// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App.vue > should render success 1`] = `""`;
